binarytrees/binarytree-fast.jl

```diff
@@ -1,59 +1,57 @@
 # The Computer Language Benchmarks Game
 # https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
 
-# contributed by Jarret Revels and Alex Arslan
-# based on an OCaml program
-# *reset*
+# contributed by Simon Danisch
+# based on the [C++ implementation](https://benchmarksgame-team.pages.debian.net/benchmarksgame/program/binarytrees-gpp-9.html)
 
 using Printf
-
-abstract type BTree end
-
-mutable struct Node <: BTree
-    left::Union{Nothing, Node}
-    right::Union{Nothing, Node}
+struct Node
+    left::Int
+    right::Int
 end
-
-function make(d)
-    if d == 0
-        Node(nothing, nothing)
-    else
-        Node(make(d-1), make(d-1))
+function alloc!(pool, left, right)
+    push!(pool, Node(left, right))
+    return length(pool)
+end
+function make(pool, d)
+    d == 0 && return 0
+    alloc!(pool, make(pool, d - 1), make(pool, d - 1))
+end
+check(pool, t::Node) = 1 + check(pool, t.left) + check(pool, t.right)
+function check(pool, node::Int)
+    node == 0 && return 1
+    @inbounds return check(pool, pool[node])
+end
+function threads_inner(pool, d, min_depth, max_depth)
+    niter = 1 << (max_depth - d + min_depth)
+    c = 0
+    for j = 1:niter
+        c += check(pool, make(pool, d))
+        empty!(pool)
     end
+    @sprintf("%i\t trees of depth %i\t check: %i\n", niter, d, c)
```

Comment:
Shouldn't we just return `niter, d, c` and do the printing after the thread loop is done? Feels safer not to do IO inside thread kernel?

Reply:
I think that would be cleaner indeed ;) I just added an `s` to printf without considering further refactoring after it passed tests!